allennlp/nn/initializers.py

```diff
@@ -198,6 +200,83 @@ def from_params(cls, params: Params):  # type: ignore
+    regardless of which parameters will actually be used to initialize the new model.
+    So, if you need to initialize several parameters using a pretrained model, the most
+    memory-efficient way to do this is to use one ``PretrainedModelInitializer`` per
+    weights file and use a regex to match all of the new parameters which need to be
+    initialized.
+
+    The below entry in the :class:`InitializerApplicator` parameters will initialize
+    ``linear_1.weight`` and ``linear_2.weight`` using a pretrained model.
+    ``linear_1.weight`` will be initialized to the pretrained
+    parameters called ``linear_1.weight``, but ``linear_2.weight`` will be initialized
+    to the pretrained parameters called ``linear_3.weight``::
+
+       ["linear_1.weight|linear_2.weight",
+           {
+               "type": "pretrained",
+               "weights_file_path": "best.th",
+               "parameter_name_overrides": {
+                   "linear_2.weight": "linear_3.weight"
+               }
+           }
+       ]
+
+    Parameters
+    ----------
+    weights_file_path : ``str``, required
+        The path to the weights file which has the pretrained model parameters.
+    parameter_name_overrides : ``Dict[str, str]``, optional (default = None)
+        The mapping from the new parameter name to the name which should be used
+        to index into the pretrained model parameters. If a parameter name is not
+        specified, the initializer will use the parameter's default name as the key.
+    """
+    def __init__(self,
+                 weights_file_path: str,
+                 parameter_name_overrides: Dict[str, str] = None) -> None:
+        self.weights: Dict[str, torch.Tensor] = torch.load(weights_file_path)
+        self.parameter_name_overrides = parameter_name_overrides or {}
+
+    @overrides
+    def __call__(self, tensor: torch.Tensor, parameter_name: str, **kwargs) -> None:  # type: ignore
+        # Select the new parameter name if it's being overridden
+        if parameter_name in self.parameter_name_overrides:
+            parameter_name = self.parameter_name_overrides[parameter_name]
+
+        # If the size of the source and destination tensors are not the
+        # same, then we need to raise an error
+        source_weights = self.weights[parameter_name]
+        if tensor.data.size() != source_weights.size():
+            raise ConfigurationError("Incompatible sizes found for parameter %s. "
+                                     "Found %s and %s" % (parameter_name,
+                                                          tensor.data.size(),
+                                                          source_weights.size()))
+
+        # Copy the parameters from the source to the destination
+        tensor.data[:] = source_weights[:]
+
+    @classmethod
+    def from_params(cls, params: Params) -> 'PretrainedModelInitializer':  # type: ignore
+        # The ``from_params`` method has to be defined, otherwise the
```

Comment:
I can remove both.

If I did not define my own `from_params`, `Initializer`'s is called, it pops the type "pretrained" and calls `PretrainedModelInitializer`'s method, which does not exist, so it actually calls itself. Then, when `cls` is `PretrainedModelInitializer`, before it pops the type from the params, it first calls `cls.list_available()`. Within this method, it checks to ensure that the default implementation is registered. Since the default implementation for `PretrainedModelInitializer` does not exist, it uses `Initializer`, which is "normal". But because this is not registered for `PretrainedModelInitializer`, it raises a ConfigurationError [here](https://github.com/allenai/allennlp/blob/193bb04ddae7d89ee63005b898a01641447d6074/allennlp/common/registrable.py#L71).

Reply:
Ah, that's a really interesting side effect of how we do `FromParams` that I hadn't considered - if we ever have a `Registrable` class that implements its own `from_params`, _all_ subclasses must also implement it.  That's unfortunate...

Reply:
I wouldn't call that a "side effect" per se -- the automatic `FromParams` is defined in the deepest base class, so once it's overridden somewhere in a class hierarchy, it's overridden the rest of the way up.

But presumably we could add logic to `Initializer.from_params` that checks if `cls == Initializer` and if not it calls `FromParams.from_params(cls, params)`, I think that would work? Maybe it's too hacky?

Reply:
I think that's a good solution for future cases where we have something like this.  For `Initializer`, it's not actually necessary, because we can just remove the method entirely.